src/serial_io.rs

```diff
@@ -0,0 +1,58 @@
+use heapless::Deque;
+use esp32c3_hal::{ 
+    Serial,
+    pac,
+    pac::UART0,
+};
+use esp_hal_common::{ 
+    serial::Instance, 
+    interrupt,
+    interrupt::*,
+    interrupt::CpuInterrupt::*,
+    Cpu::*,
+};
+use crate::protocol::InputIO;
+use crate::targets::esp32c3 as target;
+
+const RX_QUEUE_SIZE: usize = target::MAX_WRITE_BLOCK + 0x400;
+static mut RX_QUEUE: Deque<u8, RX_QUEUE_SIZE> = Deque::new();
+
+impl<'a, T: Instance> InputIO for Serial<T> {
+    fn recv(&mut self) -> u8 {
+        unsafe{
+            while riscv::interrupt::free(|_| RX_QUEUE.is_empty() ) { }
```

Comment:
This feels quite clunky. Any reason we need a deque? AFAICT we're treating the deque exactly like a queue?

If a queue is suitable, we could [split](https://docs.rs/heapless/latest/heapless/spsc/struct.Queue.html#method.split) the queue into `Producer` & `Consumer`, move the consumer here and move the producer to the interrupt.

Reply:
Initially I used `spsc::Queue`, however, the queue started to overflow after increasing baudrate to 921600. I didn't examine its implementation too thoroughly, but replacing it with `Deque` solved this issue.

Reply:
It might be that `recv` was getting starved of cpu time with many interrupts coming in. The reason why this wouldn't happen with the current setup is that the whole queue structure is locked when dequeuing, not just the `recv` part.

Reply:
I changed it back to `spsc::Queue`, just with `unchecked` version of `enqueue/dequeu` functions. Still cannot quite understand why it works this time. What has changed current setup?